python/kubeflow/trainer/utils/utils.py

```diff
@@ -107,6 +108,74 @@ def get_runtime_trainer_container(
     return None
 
 
+def _detect_trainer_from_image_patterns(image_name: str) -> Optional[types.Trainer]:
+    """
+    Detect trainer type based on image name patterns using regex.
+
+    This method uses pattern matching on the image name to determine
+    the likely trainer type.
+
+    Args:
+        image_name: The container image name
+
+    Returns:
+        Trainer object if detected, None otherwise
+    """
+    # DeepSpeed patterns
+    if re.search(r"deepspeed", image_name, re.IGNORECASE):
+        return copy.deepcopy(types.TRAINER_CONFIGS[types.Framework.DEEPSPEED])
+
+    # MLX patterns
+    if re.search(r"mlx", image_name, re.IGNORECASE):
+        return copy.deepcopy(types.TRAINER_CONFIGS[types.Framework.MLX])
+
+    # TorchTune patterns (check before PyTorch to avoid conflicts)
+    if re.search(r"torchtune", image_name, re.IGNORECASE):
+        return copy.deepcopy(types.TRAINER_CONFIGS[types.Framework.TORCHTUNE])
+
+    # PyTorch patterns (more specific to avoid matching torchtune)
+    if re.search(r"pytorch", image_name, re.IGNORECASE):
+        return copy.deepcopy(types.TRAINER_CONFIGS[types.Framework.TORCH])
+
+    # Generic torch patterns (but not torchtune)
+    if re.search(r"^torch(?!tune)", image_name, re.IGNORECASE):
+        return copy.deepcopy(types.TRAINER_CONFIGS[types.Framework.TORCH])
+
+    return None
+
+
+def _detect_trainer(
+    trainer_container: models.IoK8sApiCoreV1Container,
+) -> types.Trainer:
+    """
+    Detect trainer type with precedence logic.
+
+    This method implements the precedence order:
+    1. Check existing ALL_TRAINERS mapping (backward compatibility)
+    2. Use image pattern matching
+    3. Fall back to DEFAULT_TRAINER
+
+    Args:
+        trainer_container: The trainer container object
+
+    Returns:
+        Trainer object
+    """
+    image_name = trainer_container.image.split(":")[0]
+
+    # 1. Check existing ALL_TRAINERS mapping (backward compatibility)
+    if image_name in types.ALL_TRAINERS:
+        return copy.deepcopy(types.ALL_TRAINERS[image_name])
```

Comment:
Do we really need to maintain ALL_TRAINERS list if we are going to detect ML Framework based on regex ?

Reply:
Removed the `ALL_TRAINERS` function and the ML Framework detection is now based on the regex